ReactiveSwift-UIExamples.playground/Pages/ValidatingProperty.xcplaygroundpage/Contents.swift

```diff
@@ -0,0 +1,153 @@
+import ReactiveSwift
+import Result
+import UIKit
+import PlaygroundSupport
+
+final class ViewModel {
+	struct FormError: Error {
+		let reason: String
+
+		static let invalidEmail = FormError(reason: "The address must end with `@reactivecocoa.io`.")
+		static let mismatchEmail = FormError(reason: "The e-mail addresses do not match.")
+		static let usernameUnavailable = FormError(reason: "The username has been taken.")
+	}
+
+	let email: MutableValidatingProperty<String, FormError>
+	let emailConfirmation: MutableValidatingProperty<String, FormError>
+	let termsAccepted: MutableProperty<Bool>
+
+	let reasons: Signal<String, NoError>
+
+	let submit: Action<(), (), FormError>
+
+	init(userService: UserService) {
+		email = MutableValidatingProperty<String, FormError>("") { input, error in
+			if !input.hasSuffix("@reactivecocoa.io") {
+				error = .invalidEmail
```

Comment:
I'm not a huge fan of this pattern for returning an error. Would it not make more sense to return a `Result` here? This would also give the validator an opportunity to modify the attempted value coming in.

Code would look like:

``` swift
email = MutableValidatingProperty<String, FormError>("") { input, error in
    guard input.hasSuffix("@reactivecocoa.io") else { return .failure(.invalidEmail) }
    return .success(input)
}
```

Or, if we want to mess with the value and save the user a bit of time:

```
email = MutableValidatingProperty<String, FormError>("") { input, error in
    if input.isAlreadyPerfect {
        return .success(input)
    } else if !input.passesSomeArbitraryBadCharacterTestsAndCantBeFixed {
        return .failure(.invalidEmail)
    } else {
        let cleanInput = input.removingAllSuffixesAndWhitespace
        return .success(cleanInput + "@reactivecocoa.io")
}
```

Validation in Cocoa works in this way, where the `validateValue:forKeyPath:` call is given an in-out pointer so you can manipulate the value. I use this pattern myself in [FuzzMeasure](http://fuzzmeasure.com) when a user enters a value outside of an accepted range of values rather than barking at them with errors.

Now, it would be *really nice* if this API could somehow also allow a third, "Input was invalid but here is a new, valid value that I replaced the user's entry with" state. This ability is missing with the existing Cocoa behavior. It doesn't allow me to easily detect when the user's input was overridden so I could call attention to it in my UI. The user just sees the value flip to something different when the commit the edit on a text field right now.

But I suppose that's as easy as wrapping the `Value` itself in a new enum, so probably already possible using what you've proposed here.

For example:

``` swift
email = MutableValidatingProperty<UserInput<String>, FormError>("") { input, error in
    if input.isAlreadyPerfect {
        return .success(.unmodified(input))
    } else if !input.passesSomeArbitraryBadCharacterTestsAndCantBeFixed {
        return .failure(.invalidEmail)
    } else {
        let cleanInput = input.removingAllSuffixesAndWhitespace
        return .success(.modified(cleanInput + "@reactivecocoa.io", "Hey, we fixed that for you."))
}
```

Then, elsewhere in your code you'd just watch for those modified values and update the UI to call attention to the helpful messages.

Thoughts?